### PR TITLE
Check for duplicate string identifiers.

### PR DESCRIPTION
**Pull request by wxsBSD**
Fix an issue where duplicate string identifiers were not properly causing an
error when parsing the rules.

In the process of adding this I discovered a test I had written had in a
previous commit had duplicate string identifiers. I also discovered the test for
the base64 alphabet length was slightly incorrect in how it was doing
assertions.

### CODE REVIEW

**Comment by wxsBSD** [comment]
There are some other edge cases handled by the yara compiler that I think aren't handled in gyp that I plan to add to this PR. So far I've identified that unused strings do not cause a parser error here. I'll try and update this PR in the coming days to account for this and any other issues I'm able to tease out between the behavior of gyp and the yara compiler.

File: parser/grammar.y [hunk 1]
```diff
@@ -439,6 +439,15 @@ string_declarations
       }
     | string_declarations string_declaration
       {
+        new_ident := $2.GetIdentifier()
+        for _, ident := range $1 {
+          // Don't error if both identifiers are the empty string.
+          if old_ident := ident.GetIdentifier();  new_ident != "" && old_ident != "" && new_ident == old_ident {
+            return asLexer(yrlex).setError(
+              gyperror.DuplicateStringError,
+              `duplicate string identifier "%s"`, $2.GetIdentifier())
```
Comment (plusvic) [inline parser/grammar.y:448]: It may be useful to include the name of the rule containing the duplicate string in the error message, but I think this implies that the validation must be done at the `rule` level, when the `ast.Rule` object is being constructed. An additional complication is that in order to report the line number where the duplicate string resides, instead of the line number where the rule ends, we may need to implement a variant of `setError` that accepts the line number as an argument, instead of using the current line number. We can pass the string's line number to that variant of set error while creating the error (the `ast.String` interface has a `GetLineNo` method)

Comment (wxsBSD) [reply parser/grammar.y]: Good point! I think I've implemented it correctly but please tell me if there's a better way to do it as I'm still learning my way through the basics of golang. :)

File: parser/grammar.y [hunk 2]
```diff
@@ -280,6 +280,22 @@ rule
...
+          ident := str.GetIdentifier()
+          // Anonymous strings (no identifiers) are fine.
+          if ident == "" {
+            continue
+          }
+          if m[ident] {
+            return asLexer(yrlex).setErrorWithLineNumber(
+              gyperror.DuplicateStringError,
+              str.GetLineNo(),
+              `rule "%s" duplicate string identifier "%s"`, $<rule>4.Identifier, ident)
```
Comment (plusvic) [inline parser/grammar.y:295]: I think the message should be either...

`duplicate string identifier "foo" in rule "bar"` 

or 

`rule "bar": duplicate string identifier "foo"`

Without the colon after `rule "bar"` the message is read as a whole sentence that is a bit confusing.

**Review by plusvic — APPROVED** [review]
